sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,665 @@
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator+(size_t Offset) const;
+----
+|
+Returns an `annotated_ptr` that points to `this[Offset]`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+std::ptrdiff_t operator-(annotated_ptr FromPtr) const;
+----
+|
+Returns the distance between the underlying pointer and the pointer encapsulated
+by `FromPtr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+operator bool() const noexcept;
+----
+|
+Returns `false` if the underlying pointer is null, returns `true` otherwise.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+/*unspecified*/ operator T*() noexcept = delete;
+/*unspecified*/ operator const T*() const noexcept = delete;
+----
+|
+Implicit conversion to a pointer to the underlying type is not supported.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+T* get() noexcept;
+const T* get() const noexcept;
+----
+|
+Returns the underlying raw pointer. The raw pointer will not retain the
+annotations.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator=(const T*) noexcept;
+----
+|
+Allows assignment from a pointer to type `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator=(annotated_ptr const&) noexcept;
+----
+|
+Allows assignment from an `annotated_ptr` with the same parameterization.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator++() noexcept;
+----
+|
+Prefix increment operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator++(int) noexcept;
+----
+|
+Postfix increment operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator--() noexcept;
+----
+|
+Prefix decrement operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator--(int) noexcept;
+----
+|
+Postfix decrement operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr bool has_property();
+----
+| 
+Returns true if the property list contains the property with property key class
+`propertyT`. Returns false if it does not.
+
+Available only when `propertyT` is a property key class.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr auto get_property();
+----
+| 
+Returns a copy of the property value contained in the property list
+`PropertyListT`. Must produce a compile error if `PropertyListT` does not
+contain a property with the `propertyT` key.
+
+Available only if `propertyT` is the property key class of a compile-time
+constant property.
+|===
+
+=== Add new reference wrapper class `annotated_ref` to enable `annotated_ptr`
+
+The purpose of the `annotated_ref` class template is to provide reference
+wrapper semantics. It enables properties to be preserved on loads from and
+stores to annotated_ptrs.
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ref {
+  public:
+    annotated_ref(T *);
```

Comment:
Is there a legitimate case for users to create their own annotated_refs from a raw pointer?  If this is only for use in the SYCL headers should this constructor be private?  Marking annotated_ptr as a friend would still allow it to call the constructor.

Reply:
Removed.